### PR TITLE
[TRAFODION-37] Prepare "ROLLBACK" and "COMMIT WORK" at connection phase for T2 Rowset purpose

### PR DESCRIPTION
**Pull request by ryzuo**
Get "ROLLBACK WORK" and "COMMIT WORK" transaction prepared during the connection phase. This is for Rowsets functionality purpose, either the "COMMIT WORK" Transaction need to be executed to make it work, or "ROLLBACK WORK" shall be executed for failure.

### CODE REVIEW

File: core/conn/jdbc_type2/native/CSrvrConnect.cpp [hunk 1]
```diff
@@ -457,6 +461,23 @@ SRVR_STMT_HDL *SRVR_CONNECT_HDL::getSrvrStmt(long dialogueId,long stmtId,long
...
+
+    SRVR_STMT_HDL *pSrvrStmt = NULL;
+
+    MapOfInternalSrvrStmt::iterator iterOfStmtId = mapOfInternalSrvrStmt.find(stmtLabel);
+    if( !(iterOfStmtId == mapOfInternalSrvrStmt.end()) )
+        pSrvrStmt =(SRVR_STMT_HDL *)iterOfStmtId->second;
+
+    if(pSrvrStmt != NULL)
+        FUNCTION_RETURN_PTR(pSrvrStmt,(NULL));
+}
```
Comment (DaveBirdsall) [inline core/conn/jdbc_type2/native/CSrvrConnect.cpp]: What happens if pSrvrStmt is NULL? (I don't see a return statement for this case.)

File: core/conn/jdbc_type2/native/CSrvrConnect.h [hunk 2]
```diff
@@ -118,6 +128,8 @@ class SRVR_CONNECT_HDL
     MapOfSrvrStmt mapOfSrvrStmt;
     MapOfInternalSrvrStmt mapOfInternalSrvrStmt;
 
+	//CReplyMsgMap            replyMsgMap;
```
Comment (DaveBirdsall) [inline core/conn/jdbc_type2/native/CSrvrConnect.h]: Why are we adding this comment?

**Comment by svarnau** [comment]
jenkins, ok to test

File: core/conn/jdbc_type2/native/SqlInterface.cpp [hunk 3]
```diff
@@ -4440,3 +4440,45 @@ void CreateModulePlan(long inputParamCount, InputDescInfo *inputDescInfo, char *
...
+SQLRETURN COMMIT_ROWSET(long dialogueId, bool& bSQLMessageSet, odbc_SQLSvc_SQLError* SQLError, Int32 currentRowCount)
+{
+    SQLRETURN retcode;
+    long      sqlcode;
+    SQLValueList_def inValueList;
+    inValueList._buffer = NULL;
+    inValueList._length = 0;
+
+    SRVR_STMT_HDL *CmwSrvrStmt = getInternalSrvrStmt(dialogueId, "STMT_COMMIT_1", &sqlcode);
+    /* Should process the error here if CmwSrvrStmt is NULL */
```
Comment (DaveBirdsall) [inline core/conn/jdbc_type2/native/SqlInterface.cpp]: What's the risk in not having error processing here?

File: core/conn/jdbc_type2/native/SrvrCommon.cpp [hunk 4]
```diff
@@ -2210,3 +2266,136 @@ void print_outputValueList(SQLValueList_def *oVL, long colCount, const char * fc
...
+    msg_total_len += sizeof(tsqlState);
+
+    memcpy(MXCSWarningOrError, WarningOrError, sizeof(WarningOrError));
+    *MXCSWarningOrErrorLength = msg_total_len;
+    return;
+}
+
+bool isUTF8(const char *str)
+{
+    char c;
```
Comment (DaveBirdsall) [inline core/conn/jdbc_type2/native/SrvrCommon.cpp]: unsigned char is safer. On some platforms you'll get sign extension when comparing to literals such as 0x80, and the code won't work the way you expect.

File: core/conn/jdbc_type2/native/SrvrCommon.cpp [hunk 5]
```diff
@@ -2210,3 +2266,136 @@ void print_outputValueList(SQLValueList_def *oVL, long colCount, const char * fc
...
+    else
+        len = copySize;
+
+    if (len >= destSize)
+        len = destSize-1; // truncation
+
+    while (len > 0)
+    {
+        c = src[len-1];
+        if (c < 0x80 || c > 0xbf)
```
Comment (DaveBirdsall) [inline core/conn/jdbc_type2/native/SrvrCommon.cpp]: I'm not sure these comparisons will work the way you want with signed char.

File: core/conn/jdbc_type2/native/SrvrKds.cpp [hunk 6]
```diff
@@ -193,3 +196,43 @@ void kdsCopyToSQLValueSeq(SQLValueList_def *SQLValueList,
...
+	size_t	len;
+
+	len = strlen(msg_buf);
+	// Allocate String Buffer
+	
+	SqlErrorDesc = (ERROR_DESC_def *)SQLError->errorList._buffer + SQLError->errorList._length;
+	SqlErrorDesc->errorText = new char[len+1];
+	if (SqlErrorDesc->errorText == NULL)
+	{
+		exit(0);
```
Comment (DaveBirdsall) [inline core/conn/jdbc_type2/native/SrvrKds.cpp]: Is this what the T2 driver code normally does on a memory allocation failure? How will we know that it was a memory allocation failure?

**Comment by svarnau** [comment]
Test results are found here: https://jenkins02.trafodion.org/job/Check-PR/49/ (PASS)

File: core/conn/jdbc_type2/native/CSrvrConnect.cpp [hunk 7]
```diff
@@ -268,10 +268,10 @@ void SRVR_CONNECT_HDL::addSrvrStmt(SRVR_STMT_HDL *pSrvrStmt,BOOL internalStmt)
         mapOfInternalSrvrStmt[pSrvrStmt->stmtName]= pSrvrStmt;  // +++ map error
     }
 
-
     //End of Soln. No.: 10-100202-7923
 
     count++;
+    //pSrvrStmt->myKey = count;
```
Comment (arvind-narain) [inline core/conn/jdbc_type2/native/CSrvrConnect.cpp]: Thanks for removing some unneeded comments. Is this particular one needed ? In general if it might be good to clean up the method being touched (remove sol numbers, commented out code if not needed or add more comments, remove any MFC related code etc)

File: core/conn/jdbc_type2/native/CSrvrConnect.cpp [hunk 8]
```diff
@@ -338,7 +338,10 @@ SRVR_STMT_HDL *SRVR_CONNECT_HDL::createSrvrStmt(
...
     BOOL    useDefaultDesc,
     BOOL    internalStmt,
-    long stmtId)
+    long stmtId,
+    short sqlQueryType,      // If sqlQueryType indicates it's SPJ statement, then the member variable
+    Int32  resultSetIndex,   // resultSetIndex and callStmtId of SRVR_STMT_HDL should be initialized
+    SQLSTMT_ID* callStmtId)  // with these passed in args
 {
     FUNCTION_ENTRY("SRVR_CONNECT_HDL::createSrvrStmt",("..."));
     DEBUG_OUT(DEBUG_LEVEL_ENTRY,("  stmtLabel=%s, sqlcode=0x%08x",
```
Comment (arvind-narain) [inline core/conn/jdbc_type2/native/CSrvrConnect.cpp]: Do please add the new parameters -sqlQueryType, resultSetIndex etc.

File: core/conn/jdbc_type2/native/CSrvrConnect.cpp (hunk 1 above)
Comment (ryzuo) [inline core/conn/jdbc_type2/native/CSrvrConnect.cpp]: will fix and submit another commit within this pull request.

File: core/conn/jdbc_type2/native/SqlInterface.cpp (hunk 3 above)
Comment (ryzuo) [inline core/conn/jdbc_type2/native/SqlInterface.cpp]: For now we make sure this statement is prepared during the connection phase, MXOSRVR do it the same way without a NULL validation, I'm not sure has anyone ever met this kind of problem yet. T2 now has no mechanism to process this kind of null statement error. I'm thinking to return the error back to JAVA layer to throw an exception out, it will be included in the new rowset execute() operation code.

File: core/conn/jdbc_type2/native/SrvrCommon.cpp (hunk 4 above)
Comment (ryzuo) [inline core/conn/jdbc_type2/native/SrvrCommon.cpp]: Sorry for this, this is a whole sale copy for testing purpose, will remove it. However, strcpyUTF8() will be needed.

File: core/conn/jdbc_type2/native/SrvrKds.cpp (hunk 6 above)
Comment (ryzuo) [inline core/conn/jdbc_type2/native/SrvrKds.cpp]: T2 do has no memory allocation failure management, even in its wrapped MEMORY_ALLOC macros, it always simply exit. Memory control enhancement is another thing that we are trying to do. In the past in MXOSRVR, this kind of error is written into event repository, but now we don't have repository enabled yet, so we just keep it this way shortly.

File: core/conn/jdbc_type2/native/CSrvrConnect.cpp (hunk 7 above)
Comment (ryzuo) [inline core/conn/jdbc_type2/native/CSrvrConnect.cpp]: this key will be needed once we use byte buffer to transport between JAVA and native, and it might be used for mapping the statements. I will remove the MFC related code one by one as in this work

**Comment by svarnau** [comment]
Successful results: https://jenkins02.trafodion.org/job/Check-PR/55/

**Comment by RuoYuHP** [comment]
Will Arvind or Dave, ro anyone else please do another review?

**Comment by DaveBirdsall** [comment]
I'll take a look today. Would be good if Arvind can do so also.

**Comment by arvind-narain** [comment]
I’ll do the same today RouYu.

**Comment by DaveBirdsall** [comment]
I'm OK with the changes. If Arvind also approves, I will commit them.

**Comment by arvind-narain** [comment]
Thanks RuoYu. Looks good. Do please take care of the function entry/debug out parameters for existing and new methods in the next round of changes. Dave - ok to commit - thanks.
